snap/quota/quota.go

```diff
@@ -0,0 +1,327 @@
+	// otherwise we need to track back to get all of the parent elements
+	grpNames := []string{}
+	parentGrp := grp.parentGroup
+	for parentGrp != nil {
+		grpNames = append([]string{parentGrp.Name}, grpNames...)
+		parentGrp = parentGrp.parentGroup
+	}
+
+	buf := &bytes.Buffer{}
+	fmt.Fprintf(buf, "snap.")
+	for _, parentGrpName := range grpNames {
+		fmt.Fprintf(buf, "%s-", systemd.EscapeUnitNamePath(parentGrpName))
+	}
+	fmt.Fprintf(buf, "%s.slice", escapedGrpName)
+	return buf.String()
+}
+
+func (grp *Group) validate() error {
+	if err := naming.ValidateQuotaGroup(grp.Name); err != nil {
+		return err
+	}
+
+	// check if the name is reserved for future usage
+	switch grp.Name {
+	case "root", "system", "snapd", "user":
+		return fmt.Errorf("group name %q reserved", grp.Name)
+	}
+
+	if grp.MemoryLimit == 0 {
+		return fmt.Errorf("group memory limit must be non-zero")
+	}
+
+	// TODO: probably there is a minimum amount of bytes here that is
+	// technically usable/enforcable, should we check that too?
+
+	if grp.ParentGroup != "" && grp.Name == grp.ParentGroup {
+		return fmt.Errorf("group has circular parent reference to itself")
+	}
+
+	if len(grp.SubGroups) != 0 {
+		for _, subGrp := range grp.SubGroups {
+			if subGrp == grp.Name {
+				return fmt.Errorf("group has circular sub-group reference to itself")
+			}
+		}
+	}
+
+	// check that if this is a sub-group, then the parent group has enough space
+	// to accommodate this new group (we assume that other existing sub-groups
+	// in the parent group have already been validated)
+	if grp.parentGroup != nil {
+		alreadyUsed := quantity.Size(0)
+		for _, child := range grp.parentGroup.subGroups {
+			if child.Name == grp.Name {
+				continue
+			}
+			alreadyUsed += child.MemoryLimit
+		}
+		// careful arithmetic here in case we somehow overflow the max size of
+		// quantity.Size
+		if grp.parentGroup.MemoryLimit-alreadyUsed < grp.MemoryLimit {
+			remaining := grp.parentGroup.MemoryLimit - alreadyUsed
+			return fmt.Errorf("sub-group memory limit of %s is too large to fit inside remaining quota space %s for parent group %s", grp.MemoryLimit.IECString(), remaining.IECString(), grp.parentGroup.Name)
+		}
+	}
+
+	return nil
+}
+
+// NewSubGroup creates a new sub group under the current group.
+func (grp *Group) NewSubGroup(name string, memLimit quantity.Size) (*Group, error) {
+	subGrp := &Group{
+		Name:        name,
+		MemoryLimit: memLimit,
+		ParentGroup: grp.Name,
+		parentGroup: grp,
+	}
+
+	// check early that the sub group name is not the same as that of the
+	// parent, this is fine in systemd world, but in snapd we want unique quota
+	// groups
+	if name == grp.Name {
+		return nil, fmt.Errorf("cannot use same name %q for sub group as parent group", name)
+	}
+
+	if err := subGrp.validate(); err != nil {
+		return nil, err
+	}
+
+	// save the details of this new sub-group in the parent group
+	grp.subGroups = append(grp.subGroups, subGrp)
+	grp.SubGroups = append(grp.SubGroups, name)
```

Comment:
Nitpick, I found subGroups and SubGroups slightly confusing when I reached this line, maybe one should be renamed to SubGroupNames, or the non-exported one to just `groups` (since it's a member under Group struct, it's in a way obvious it is about sub-groups. But not a blocker.